src/analysis_and_optimization/Debug_data_generation.ml

```diff
@@ -274,27 +274,34 @@ and generate_value m st t =
       gen_complex_matrix (unwrap_int_exn m e1) (unwrap_int_exn m e2)
   | SArray (st, e) -> gen_array m st (unwrap_int_exn m e) t
 
-let rec pp_value_json ppf e =
-  match e.Expr.Fixed.pattern with
-  | Lit (Real, s) when String.is_suffix s ~suffix:"." -> Fmt.string ppf (s ^ "0")
-  | Lit ((Int | Real), s) -> Fmt.string ppf s
-  | FunApp (CompilerInternal (FnMakeRowVec | FnMakeArray), l) ->
-      Fmt.(pf ppf "[@[<hov 1>%a@]]" (list ~sep:comma pp_value_json) l)
-  | FunApp (StanLib (transpose, _, _), [e])
-    when String.equal transpose (Operator.to_string Transpose) ->
-      pp_value_json ppf e
-  | FunApp (StanLib ("to_complex", _, _), [r; i]) ->
-      Fmt.(pf ppf "[@[<hov 1>%a, %a@]]" pp_value_json r pp_value_json i)
-  | _ ->
-      Common.FatalError.fatal_error_msg
-        [%message "Could not evaluate expression " (e : Expr.Typed.t)]
-
-let print_data_prog s =
-  let l, _ =
-    List.fold s ~init:([], Map.Poly.empty)
-      ~f:(fun (l, m) (sizedtype, transformation, name) ->
-        let value = generate_value m sizedtype transformation in
-        ((name, value) :: l, Map.set m ~key:name ~data:value) ) in
-  let pp ppf (id, value) =
-    Fmt.pf ppf {|@[<hov 2>"%s":@ %a@]|} id pp_value_json value in
-  Fmt.(str "{@ @[<hov>%a@]@ }" (list ~sep:comma pp) (List.rev l))
+let generate_expressions data =
+  List.fold data ~init:([], Map.Poly.empty)
+    ~f:(fun (l, m) (sizedtype, transformation, name) ->
+      let value = generate_value m sizedtype transformation in
+      ((name, value) :: l, Map.set m ~key:name ~data:value) )
+  |> fst |> List.rev
+
+open Yojson
+
+let generate_json_entries (name, expr) : string * t =
+  let rec expr_to_json e : t =
+    match e.Expr.Fixed.pattern with
+    | Lit (Real, s) when String.is_suffix s ~suffix:"." -> `Floatlit (s ^ "0")
```

Comment:
This suffix fudging is kind of ugly. Changing this to `Float (Float.of_string s)` seems to lose one decimal, not sure if that matters (and not sure if it's any less of a hack...)

Reply:
I don't think JSON makes any distinction between `0`, `0.`, and `0.0`, does it?

The best way of doing this would be to avoid turning the float we generate higher up in the file into a string in the first place, but that's not really easy to do with the way Expressions are set up. 

We could also change `Float.to_string` to something like `Float.to_string_hum ~decimals:16 ~strip_zero:false`, which will generate `0.0000000000000000` instead, though obviously it would be great if there was a 'strip all but one' zeros option

Reply:
Alternatively, we could move this match logic into `Expr.Helpers.float`

Reply:
I think `0` and `0.0` are the same but `0.` is a syntax error. I don't like `0.0000000000000000`. This isn't really worth overengineering so let's leave it as a suffix check for now.